openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1577 @@
+        //dilateVoxels(mgr, dilation, nn);
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.restart("Initializing grid and sweep mask");
+#endif
+
+        mParent->mSweepMask.clear();
+        mParent->mSweepMask.topologyUnion(mParent->mSdfGrid->constTree());
+
+        using LeafManagerT = tree::LeafManager<typename SdfGridT::TreeType>;
+        using LeafT = typename SdfGridT::TreeType::LeafNodeType;
+        LeafManagerT leafManager(mParent->mSdfGrid->tree());
+
+        auto kernel = [&](LeafT& leaf, size_t /*leafIdx*/) {
+            static const SdfValueT Unknown = std::numeric_limits<SdfValueT>::max();
+            const SdfValueT background = mBackground;//local copy
+            auto* maskLeaf = mParent->mSweepMask.probeLeaf(leaf.origin());
+            assert(maskLeaf);
+            for (auto voxelIter = leaf.beginValueOn(); voxelIter; ++voxelIter) {
+                const SdfValueT value = *voxelIter;
+                if (math::Abs(value) < background) {// disable boundary voxels from the mask tree
+                    maskLeaf->setValueOff(voxelIter.pos());
+                } else {
+                    voxelIter.setValue(value > 0 ? Unknown : -Unknown);
+                }
+            }
+        };
+
+        leafManager.foreach( kernel );
+
+        // cache the leaf node origins for fast lookup in the sweeping kernels
+
+        mParent->computeSweepMaskLeafOrigins();
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        timer.stop();
+#endif
+    }// FastSweeping::DilateKernel::run
+
+    // Private member data of DilateKernel
+    FastSweeping   *mParent;
+    const SdfValueT  mBackground;
+};// FastSweeping::DilateKernel
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename SdfGridT, typename ExtValueT>
+struct FastSweeping<SdfGridT, ExtValueT>::InitSdf
+{
+    using LeafRange = typename tree::LeafManager<SdfTreeT>::LeafRange;
+    InitSdf(FastSweeping &parent): mParent(&parent),
+      mSdfGrid(parent.mSdfGrid.get()), mIsoValue(0), mAboveSign(0) {}
+    InitSdf(const InitSdf&) = default;// for tbb::parallel_for
+    InitSdf& operator=(const InitSdf&) = delete;
+
+    void run(SdfValueT isoValue, bool isInputSdf)
+    {
+        mIsoValue   = isoValue;
+        mAboveSign  = isInputSdf ? SdfValueT(1) : SdfValueT(-1);
+        SdfTreeT &tree = mSdfGrid->tree();//sdf
+        const bool hasActiveTiles = tree.hasActiveTiles();
+
+        if (isInputSdf && hasActiveTiles) {
+          OPENVDB_THROW(RuntimeError, "FastSweeping: A SDF should not have active tiles!");
+        }
+
+#ifdef BENCHMARK_FAST_SWEEPING
+        util::CpuTimer  timer("Initialize voxels");
+#endif
+        mParent->mSweepMask.clear();
+        mParent->mSweepMask.topologyUnion(mParent->mSdfGrid->constTree());
+
+        {// Process all voxels
+          tree::LeafManager<SdfTreeT> mgr(tree, 1);// we need one auxiliary buffer
+          tbb::parallel_for(mgr.leafRange(32), *this);//multi-threaded
```

Comment:
Still manually specifying a leaf range here of 32 per thread. Is this intentional?

Reply:
see comment above - this is indeed intentional